rust-multiaddr/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use multiaddr::Multiaddr;
+    ///
+    /// let address = Multiaddr::new("/ip4/127.0.0.1").unwrap();
+    /// let nested = address.encapsulate("/udt").unwrap();
+    /// assert_eq!(nested, Multiaddr::new("/ip4/127.0.0.1/udt").unwrap());
+    /// ```
+    ///
+    pub fn encapsulate<T: ToMultiaddr>(&self, input: T) -> Result<Multiaddr> {
+        let new = input.to_multiaddr()?;
+        let mut bytes = self.bytes.clone();
+
+        bytes.extend(new.to_bytes());
+
+        Ok(Multiaddr { bytes: bytes })
+    }
+
+    /// Remove the outer most address from itself.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use multiaddr::{Multiaddr, ToMultiaddr};
+    ///
+    /// let address = Multiaddr::new("/ip4/127.0.0.1/udt/sctp/5678").unwrap();
+    /// let unwrapped = address.decapsulate("/udt").unwrap();
+    /// assert_eq!(unwrapped, Multiaddr::new("/ip4/127.0.0.1").unwrap());
+    ///
+    /// assert_eq!(
+    ///     address.decapsulate("/udt").unwrap(),
+    ///     "/ip4/127.0.0.1".to_multiaddr().unwrap()
+    /// );
+    /// ```
+    ///
+    /// Returns the original if the passed in address is not found
+    ///
+    /// ```
+    /// use multiaddr::ToMultiaddr;
+    ///
+    /// let address = "/ip4/127.0.0.1/udt/sctp/5678".to_multiaddr().unwrap();
+    /// let unwrapped = address.decapsulate("/ip4/127.0.1.1").unwrap();
+    /// assert_eq!(unwrapped, address);
+    /// ```
+    ///
+    pub fn decapsulate<T: ToMultiaddr>(&self, input: T) -> Result<Multiaddr> {
+        let input = input.to_multiaddr()?.to_bytes();
+
+        let bytes_len = self.bytes.len();
+        let input_length = input.len();
+
+        let mut input_pos = 0;
+        let mut matches = false;
+
+        for (i, _) in self.bytes.iter().enumerate() {
+            let next = i + input_length;
+
+            if next > bytes_len {
+                continue;
+            }
+
+            if &self.bytes[i..next] == input.as_slice() {
+                matches = true;
+                input_pos = i;
+                break;
+            }
+        }
+
+        if !matches {
+            return Ok(Multiaddr { bytes: self.bytes.clone() });
+        }
+
+        let mut bytes = self.bytes.clone();
+        bytes.truncate(input_pos);
+
+        Ok(Multiaddr { bytes: bytes })
+    }
+
+    /// Returns the components of this multiaddress.
+    /// 
+    /// ```
+    /// use std::net::Ipv4Addr;
+    /// use multiaddr::AddrComponent;
+    /// use multiaddr::Multiaddr;
+    ///
+    /// let address: Multiaddr = "/ip4/127.0.0.1/udt/sctp/5678".parse().unwrap();
+    /// 
+    /// let components = address.iter().collect::<Vec<_>>();
+    /// assert_eq!(components[0], AddrComponent::IP4(Ipv4Addr::new(127, 0, 0, 1)));
+    /// assert_eq!(components[1], AddrComponent::UDT);
+    /// assert_eq!(components[2], AddrComponent::SCTP(5678));
+    /// ```
+    ///
+    #[inline]
+    pub fn iter(&self) -> Iter {
```

Comment:
If we have this method then we should implement `IntoIterator` for `&Multiaddr`

Reply:
@tomaka Same as above, either comment why this is fine or address it.

Reply:
I made the change.